src/NuGet.Core/NuGet.Protocol/Utility/CachingUtility.cs

```diff
@@ -98,5 +99,30 @@ public static string RemoveInvalidFileNameChars(string value)
                 .Replace("__", "_");
 #endif
         }
+ 
+        private static string ToHex(byte[] bytes, int length)
```

Comment:
Rather than copy-and-paste code, how about creating in NuGet.Common:
```C#
public static class EncodingUtility
{
    public static string ToHex(...) { ... }
}
```

Reply:
There's effectively zero risk that business requirements will change that force us to change the implementation and therefore bugs existing because different implementations of the method have different outputs. The most common concern about code duplication is about bugs, so if there's very little risk, I'm not sure what the benefit is.

We've started trying to be more mindful about the [NuGet SDK](https://github.com/NuGet/NuGet.Client/blob/dev/docs/nuget-sdk.md). Since we have no way of knowing when a customer is using the SDK, let alone which APIs are being used, once an API is shipped, we should aim for keeping backwards compatibility. While this is an argument for allowing `ToHex`, because it's unlikely to ever change and therefore break customers using it, `ToHex` is not a NuGet feature. In the scope of what APIs you would expect NuGet nupkgs to provide, `ToHex` is not one of them. A customer looking for a `ToHex` method would be unreasonable expect it to be provided by the NuGet SDK.

If anyone feels strongly about removing code "duplication", I'd suggest considering our [shared code](https://github.com/NuGet/NuGet.Client/tree/dev/build/Shared), rather than a public API. The length of this comment is not indicative of my objection. Since `ToHex` is unlikely to ever change, I won't be upset if it gets added, I'd just prefer it wasn't a public API. I just want us to be more mindful and deliberate about what public APIs get added to the NuGet SDK.

Reply:
Good points.  I forgot about the recent NuGet SDK.

However, we need to be able to share implementations for internal use.  We have multiple places that can benefit from a `ToHex(...)` utility method.  Two here, but there are more elsewhere (search for `BitConverter`).

The bugs need not be functional.  They can be performance-related.  Having multiple implementations scattered throughout the code base represent maintenance overhead which we don't need.

I like the shared code suggestion.

Reply:
I've not been following this discussion in depth, and I don't intend to catch-up to everything, but I am definitely in favor of shared code for utility methods over NuGet.Common. 

To a reasonable degree ofc. :)

Reply:
Thanks! Added it to NuGet.Shared.